src/ripple/app/consensus/RCLValidations.cpp

```diff
@@ -240,24 +240,45 @@ RCLValidations::onStale(RCLValidation&& v)
         return;
```

Comment:
I'm worried about `RCLValidations::onStale()`.  In the old implementation the moral equivalent of `onStale()` was always called with `RCLValidations::lock_` held by a `std::lock_guard`.  Now all of the calls to `onStale` are coming from outside `RCLValidations`.  So I'm fairly confident the callers don't hold the `lock_`.

Any data members of `RCLValidations` that are accessed by `RCLValidations::doWrite()` need to always be protected by `lock_`.  That would include both `staleValidations_` and `writing_`.

It may be sufficient to simply grab the `lock_` at the very top of `onStale()`.  I'm unsure because I don't know the requirements of all of the places that call `onStale()`.  The result might look like this:
```
    {
        ScopedLockType sl(lock_);
        staleValidations_.emplace_back(std::move(v.val_));
        if (writing_)
            return;

        writing_ = true;
    }
    app_.getJobQueue().addJob(...
```
Once the locking situation is cleared up I'll look hard at the code again.